generator.py

```diff
@@ -0,0 +1,180 @@
+import argparse
+import re
+from collections import defaultdict
+from os import chmod, makedirs, mkdir, symlink
+from pathlib import Path
+from shutil import copy2, rmtree
+from subprocess import CalledProcessError, check_output
+
+import yaml
+from distgen.multispec import Multispec
+
+
+def run_distgen(src, dest, multispec_path, distro_config, version):
+    cmd = [
+        "dg",
+        "--multispec",
+        multispec_path,
+        "--template",
+        src,
+        "--distro",
+        distro_config,
+        "--multispec-selector",
+        f"version={version}",
+        "--output",
+        str(dest),
+    ]
+
+    try:
+        check_output(cmd)
+    except CalledProcessError as e:
+        print("[ERROR] distgen failed:", e)
+
+
+def get_version_distro_mapping(multispec_file):
+    """Get all combinations from multispec file like:
+
+    [{"distro": "rhel-8-x86_64.yaml", "version": "3.9"},
+     {"distro": "rhel-8-x86_64.yaml", "version": "3.8"},
+     {"distro": "centos-7-x86_64.yaml", "version": "3.8"},
+     {"distro": "centos-stream-9-x86_64.yaml", "version": "3.9"}]
+
+    and transfer them to:
+
+    {"3.8": ["rhel-8-x86_64.yaml", "centos-7-x86_64.yaml"],
+     "3.9": ["rhel-8-x86_64.yaml", "centos-stream-9-x86_64.yaml"]}
+    """
+    multispec_yaml = yaml.load(multispec_file.read(), Loader=yaml.SafeLoader)
+    multispec = Multispec(data=multispec_yaml)
+    mapping = defaultdict(list)
+    for combination in multispec.get_all_combinations():
+        mapping[combination["version"]].append(combination["distro"])
+    return mapping
+
+
+def filename_to_distro_config(filename, version, mapping):
+    """Find distgen distro config from a filename.
+
+    This is usually needed only for dockerfiles.
+    - Dockerfile.rhelXX → rhel-XX-x86_64.yaml
+    - Dockerfile.cXXs → centos-stream-XX-x86_64.yaml
+    - Dockerfile.fedora → the newest fedora-XX-x86_64.yaml
+    - Dockerfile → centos-7-x86_64.yaml
```

Comment:
Food for thought: "centos-7" might not be a good default anymore. How about creating a suffix for centos 7 (e.g. `centos7`) and error out when no suffix is given?

Reply:
Done. It requires handling of dead symlinks but that isn't complicated.